### DISCUSSION TITLE
Question: Handling Linux symlinks on a dual-boot setup

### DISCUSSION THREAD

**Discussion by gctypo** [abraunegg/onedrive#1652] [Q&A] [answered]
I'm dual-booting EndeavourOS with Windows 10. I have it set up such that I have a "shared" partition where I store documents/music/projects/etc - platform-independent files. Most of these files are with my OneDrive folder. I still use Windows from time to time, and while it isn't my main OS anymore, I do use it in segments upwards of several weeks sometimes. I have Microsoft's OneDrive client set up to that folder on Windows, and am trying to set up abraunegg/onedrive for that folder on Linux.

Due to where certain projects are placed, and how I frequently work with files from different locations within this OneDrive folder, I've been setting up symlinks between them. The thing is, symlinks in Linux are different than symlinks in Windows, and are definitely different than Windows shortcuts. MS's OneDrive doesn't exactly provide a lot of options, so if it sees a Linux symlink I've set up it just sees it as some extensionless file and uploads it as an extensionless file. But when I boot back to Linux and run a sync in abraunegg/onedrive, it sees that extensionless file in cloud storage, attempts to download it, and throws an error because it conflicts with what it sees as a *directory* of the same name, since it sees the symlink and tries to dereference it for contents inside.

I seem to have a few options:
1. Rely entirely on abraunegg/onedrive and stop using MS OneDrive for syncing, and turn off symlink uploads.
2. Ignore the errors and hope no shenanigans occur regarding the symlinks and conflicts that arise.
3. Stop using Linux symlinks within OneDrive, entirely.

Option 1 seems like the least problematic option, but if there was a way to get abraunegg/onedrive to not dereference symlinks and instead treat them as plain files (much like MS OneDrive treats Windows LNK shortcuts), that seems like it would alleviate this problem entirely. Is this at all possible?

---

**Setup:**

*onedrive-abraunegg 2.4.13-1*
EndeavourOS, kernel 5.14.9-arch2-1, zsh 5.8, KDE Plasma 5.22.5

**Comment by abraunegg** [comment, marked as answer]
@gctypo 
> Option 1 seems like the least problematic option, but if there was a way to get abraunegg/onedrive to not dereference symlinks and instead treat them as plain files (much like MS OneDrive treats Windows LNK shortcuts), that seems like it would alleviate this problem entirely. Is this at all possible?

When the client currently 'follows' a symbolic link under Linux, it creates the files | folders on OneDrive as per how the file system looks, when you are following those symbolic links on the Linux filesystem.

They should not be uploaded as any sort of other file to OneDrive when running under Linux.

**Example:**
```
total 8
drwxrwxr-x.  6 alex alex  144 Oct  9 12:26 .
drwx------. 40 alex alex 4096 Oct  9 12:26 ..
drwxrwxr-x.  6 alex alex   62 Oct  9 12:26 Documents
drwxrwxr-x.  3 alex alex   28 Oct  9 12:26 Dropbox
-rw-rw-r--.  1 alex alex   25 Oct  9 12:26 file1.tmp
drwxrwxr-x.  2 alex alex   26 Oct  9 12:26 Images
drwxrwxr-x.  2 alex alex   26 Oct  9 12:26 Music
lrwxrwxrwx.  1 alex alex   32 Oct  9 12:26 random_files -> /home/alex/OneDrive/random_files
lrwxrwxrwx.  1 alex alex   33 Oct  9 12:26 random_images -> /home/alex/OneDrive/random_images
lrwxrwxrwx.  1 alex alex   33 Oct  9 12:26 random_videos -> /home/alex/OneDrive/random_videos
```

When the 'random_*' folders are uploaded, they are not uploaded as a link, but an actual directory on OneDrive:

```
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Updated Remaining Free Space: 5368709120
Uploading differences of ~/OneDrivePersonal
Processing .
The directory has not changed
Uploading new items of ~/OneDrivePersonal
OneDrive Client requested to create remote path: ./random_files
The requested directory to create was not found on OneDrive - creating remote directory: ./random_files
Successfully created the remote directory ./random_files on OneDrive
OneDrive Client requested to create remote path: ./random_images
The requested directory to create was not found on OneDrive - creating remote directory: ./random_images
Successfully created the remote directory ./random_images on OneDrive
OneDrive Client requested to create remote path: ./random_images/Q4NvqmP1DdO2syUQCOxWoC4CAGx27LH3
The requested directory to create was not found on OneDrive - creating remote directory: ./random_images/Q4NvqmP1DdO2syUQCOxWoC4CAGx27LH3
Successfully created the remote directory ./random_images/Q4NvqmP1DdO2syUQCOxWoC4CAGx27LH3 on OneDrive
Uploading new file ./random_images/Q4NvqmP1DdO2syUQCOxWoC4CAGx27LH3/image0.png ... done.
Remaining free space on OneDrive: 5368477888
Uploading new file ./random_images/Q4NvqmP1DdO2syUQCOxWoC4CAGx27LH3/image1.png ... done.
Remaining free space on OneDrive: 5368246656

```

On OneDrive online, these are now 'folders' - not links | shortcuts | symbolic links:

![Capture](https://user-images.githubusercontent.com/4956234/136639204-998ceb0a-546c-45ce-89b6-8954bbc1d9cc.JPG)

On a Windows PC now, these files | folders from OneDrive are downloaded as full items.

However - if you are now sharing the local filesystem between Linux | Windows, Windows will see the Linux symbolic links as something else, thus this is where your mess is coming in from .. not the Linux side.

The only suggestions I can provide are:

1. Dont use symbolic links to link your data
2. Use symbolic links, but use `--skip-symlinks` and skip syncing of that data
3. Use 'on-demand' OneDrive functionality in Windows & Linux. To do so in Linux you need to use https://github.com/jstaf/onedriver rather than this client
4. Use separate locations for Linux and Windows for your data, then use a local 'rsync' process to sync your locations. You may need to use a filesystem that utilises de-dupe capabilities to ensure that you do not have duplicate files.
